### PR TITLE
Simple distributed optimizer

### PR DESCRIPTION
**Pull request by aazzolini**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29304 Simple distributed optimizer**

Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18354586](https://our.internmc.facebook.com/intern/diff/D18354586/)

### CODE REVIEW

**Comment by aazzolini** [comment]
This PR replaces  https://github.com/pytorch/pytorch/pull/28910 which was closed by mistake and couldn't be reopened.

For this version, I got rid of FunctionalOptimizer and made torch.optim.Optimizer work directly with DistributedOptimizer, simplifying the implementation and reducing new API footprint.

File: torch/distributed/optim/optimizer.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    Args:
+        optimizer_class (FunctionalOptimizer): the class of optimizer to
+            instantiate on each worker.
+        params_rref (list[RRef]): list of RRefs to local or remote parameters
+            to optimize.
+        args: arguments to pass to the optimizer constructor on each worker.
+        kwargs: arguments to pass to the optimizer constructor on each worker.
+    """
+    def __init__(self, optimizer_class, params_rref, *args, **kwargs):
+        per_worker_params_rref = defaultdict(lambda: [])
```
Comment (zzzwen) [inline torch/distributed/optim/optimizer.py]: nit: `defaultdict(list)`

**Review by pietern — COMMENTED** [review]
Nice! This is very simple indeed :)

File: torch/distributed/optim/optimizer.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+from threading import Lock
+
+
+class _LocalOptimizer:
+    def __init__(self, optim_cls, local_params_rref, *args, **kwargs):
+        self.optim = optim_cls(
+            [rref.local_value().wait() for rref in local_params_rref],
+            *args,
+            **kwargs)
+        self.lock = Lock()
```
Comment (vincentqb) [inline torch/distributed/optim/optimizer.py]: This is the lock mentioned [here](https://github.com/pytorch/pytorch/pull/28910#discussion_r341277257) to, right? How would we modify this to get Hogwild?

**Comment by albanD** [comment]
What is the plan for adding a tutorial / example?
Are you planning on adding this here? Or wait to have all the features to do a single tutorial with a full training?

**Comment by mrshenli** [comment]
> What is the plan for adding a tutorial / example?

We need thorough docstrings for this one, then we can have a single tutorial for full training I think.

File: torch/distributed/optim/optimizer.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    def step(self, autograd_ctx_id):
+        all_local_grads = dist_autograd.get_gradients(autograd_ctx_id)
+
+        with self.lock:
+            for param, grad in all_local_grads.items():
+                param.grad = grad
+            self.optim.step()
```
Comment (mrshenli) [inline torch/distributed/optim/optimizer.py]: this is nice!

File: torch/distributed/optim/optimizer.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    This class uses distributed autograd in order to retrieve the gradients for
+    specific parameters.
+
+    Args:
+        optimizer_class (FunctionalOptimizer): the class of optimizer to
+            instantiate on each worker.
+        params_rref (list[RRef]): list of RRefs to local or remote parameters
+            to optimize.
+        args: arguments to pass to the optimizer constructor on each worker.
+        kwargs: arguments to pass to the optimizer constructor on each worker.
```
Comment (mrshenli) [inline torch/distributed/optim/optimizer.py]: Shall we add an example here?

File: torch/distributed/optim/optimizer.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+class DistributedOptimizer:
+    """
+    DistributedOptimizer takes remote references to parameters scattered
+    across workers and applies the given optimizer locally for each parameter.
+
+    This class uses distributed autograd in order to retrieve the gradients for
+    specific parameters.
+
+    Args:
+        optimizer_class (FunctionalOptimizer): the class of optimizer to
```
Comment (mrshenli) [inline torch/distributed/optim/optimizer.py]: There is no `FunctionalOptimizer` any more, do you mean `optim.Optimizer`?

File: torch/distributed/optim/optimizer.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            instantiate on each worker.
+        params_rref (list[RRef]): list of RRefs to local or remote parameters
+            to optimize.
+        args: arguments to pass to the optimizer constructor on each worker.
+        kwargs: arguments to pass to the optimizer constructor on each worker.
+    """
```
Comment (mrshenli) [inline torch/distributed/optim/optimizer.py]: Do we want to add a warning that we only make sure concurrent `step()` calls will not modify the same param at the same time, but they could still modify params on different owners in an interleaving way? This means that when an dist optimizer tries to apply a grad x to a param, the param might already be different from when grad x was computed. This behavior is by design. If the application needs to do global exclusive dist optimizer `step()`, they will have to synchronize it on their own.

**Review by mrshenli — APPROVED** [review]
This LGTM! My comments are mostly on docs.

File: torch/distributed/optim/optimizer.py (hunk 2 above)
Comment (aazzolini) [reply torch/distributed/optim/optimizer.py]: Yes. There's no simple way of modifying the current implementation to get hogwild. We'd have to either 1) change the interface (FunctionalOptimizer); or some alternative such as keeping an object pool / or thread local instances of optim.Optimizers to avoid gradient sharing across threads. I can write a comment on why the lock is there.

File: torch/distributed/optim/optimizer.py (hunk 4 above)
Comment (aazzolini) [reply torch/distributed/optim/optimizer.py]: I think that a convincing example will need a way to 1) create the remote module; 2) call it; 3) get a list of RRef params for it. We can add this after we have closed on a way of doing these.

File: torch/distributed/optim/optimizer.py (hunk 2 above)
Comment (vincentqb) [reply torch/distributed/optim/optimizer.py]: Cool, just wanted to verify. Thanks!

**Review by vincentqb — APPROVED** [review]

**Review by mrshenli — CHANGES_REQUESTED** [review]
Internal test failure is real

File: torch/distributed/optim/optimizer.py (hunk 2 above)
Comment (pritamdamania87) [reply torch/distributed/optim/optimizer.py]: Wouldn't removing this lock give us some form of hogwild? I'm not sure how accurate this example is: https://pytorch.org/docs/stable/notes/multiprocessing.html#hogwild, although it does something similar where accumulating gradients on the tensors and running the optimizer could interleave among different processes.

**Review by mrshenli — APPROVED** [review]
Test all passed, let's land this!
